app/controllers/floating_ip_controller.rb

```diff
@@ -12,4 +12,275 @@ class FloatingIpController < ApplicationController
   def self.display_methods
     %w()
   end
+
+  def button
+    @edit = session[:edit] # Restore @edit for adv search box
+    params[:display] = @display if %w(vms instances images).include?(@display)
+    params[:page] = @current_page unless @current_page.nil? # Save current page for list refresh
+
+    @refresh_div = "main_div"
+    return tag("FloatingIp") if params[:pressed] == "floating_ip_tag"
+    delete_floating_ips if params[:pressed] == 'floating_ip_delete'
+
+    if params[:pressed] == "floating_ip_edit"
+      checked_floating_ip_id = get_checked_floating_ip_id(params)
+      javascript_redirect :action => "edit", :id => checked_floating_ip_id
+    elsif params[:pressed] == "floating_ip_new"
+      javascript_redirect :action => "new"
+    elsif !flash_errors? && @refresh_div == "main_div" && @lastaction == "show_list"
+      replace_gtl_main_div
+    else
+      render_flash
+    end
+  end
+
+  def cancel_action(message)
+    session[:edit] = nil
+    @breadcrumbs.pop if @breadcrumbs
+    javascript_redirect :action    => @lastaction,
+                        :id        => @floating_ip.id,
+                        :display   => session[:floating_ip_display],
+                        :flash_msg => message
+  end
+
+  def create
+    assert_privileges("floating_ip_new")
+    case params[:button]
+    when "cancel"
+      javascript_redirect :action    => 'show_list',
+                          :flash_msg => _("Add of new %{model} was cancelled by the user") % {
+                            :model => ui_lookup(:table => 'floating_ip') }
+
+    when "add"
+      @floating_ip = FloatingIp.new
+      options = form_params
+      ems = ExtManagementSystem.find(options[:ems_id])
+      options.delete(:ems_id)
+      task_id = ems.create_floating_ip_queue(session[:userid], options)
+
+      add_flash(_("Floating IP creation failed: Task start failed: ID [%{id}]") %
+      {:id => task_id.inspect}, :error) unless task_id.kind_of?(Fixnum)
+
+      if @flash_array
+        javascript_flash(:spinner_off => true)
+      else
+        initiate_wait_for_task(:task_id => task_id, :action => "create_finished")
+      end
+    end
+  end
+
+  def create_finished
+    task_id = session[:async][:params][:task_id]
+    floating_ip_name = session[:async][:params][:name]
+    task = MiqTask.find(task_id)
+    if MiqTask.status_ok?(task.status)
+      add_flash(_("%{model} \"%{name}\" created") % { :model => ui_lookup(:table => 'floating_ip'),
+                                                      :name  => floating_ip_name })
```

Comment:
To make life of our translators easier + to make the translations look better, the flash string here should
read `"%Floating IP \"%{name}\" created"`. Same in other places.